static/src/javascripts-legacy/projects/common/utils/check-mediator.js

```diff
@@ -0,0 +1,146 @@
+                list:[{
+                    id: 'isUserInAClashingAbTest'
+                }, {
+                    id: 'canEmailBeInserted',
+                    dependentChecks: {
+                        passCondition: EVERYCHECKPASSED,
+                        list: [{
+                            id: 'emailCanRun'
+                        }, {
+                            id: 'listCanRun'
+                        }, {
+                            id: 'emailInArticleOutbrainEnabled'
+                        }]
+                    }
+                }]
+            }
+        }];
+
+    function DeferredCheck(dependentCheckPromises, dependentChecksPassCondition) {
+        this.complete = new Promise(function(resolve, reject) {
+            this.resolve = resolve;
+            this.reject = reject;
+        }.bind(this));
+
+        if (dependentCheckPromises) {
+            Promise.all(dependentCheckPromises).then(function(results) {
+
+                var hasPassed = function(result) {
+                    return result;
+                };
+
+                if (dependentChecksPassCondition.call(results, hasPassed)) {
+                    this.resolve(true);
+                } else {
+                    this.resolve(false);
+                }
+            }.bind(this));
+        }
+    }
+
+    function registerDefferedCheck(check) {
+        var dependentCheckPromises = [];
+
+        if (check.dependentChecks) {
+            check.dependentChecks.list.forEach(registerDependentCheck.bind(null, dependentCheckPromises));
+        
+            return new DeferredCheck(dependentCheckPromises, check.dependentChecks.passCondition);
+        }
+
+        return new DeferredCheck();
+    }
+
+    
+    function registerDependentCheck(dependentCheckPromises, dependentCheck) {
+        dependentCheckPromises.push(registerCheck(dependentCheck).complete);
+    }    
+
+    function registerCheck(check) {
+        var registeredCheck = registerDefferedCheck(check);
+
+        registeredChecks[check.id] = registeredCheck;
+
+        return registeredCheck;
+    }
+
+    /**
+     * public
+    **/
+    function init() {
+        checkList.forEach(registerCheck);
+    }
+    
+    function resolveCheck(id) {
+        var argsArray = Array.prototype.slice.call(arguments, 1);
+        
+        if (registeredChecks[id]) {
+            return registeredChecks[id].resolve.apply(null, argsArray);
+        }
+    }
+
+    function rejectCheck(id) {
+        var argsArray = Array.prototype.slice.call(arguments, 1);
+
+        if (registeredChecks[id]) {
+            return registeredChecks[id].reject.apply(null, argsArray);
+        }
+    }
+
+    function waitForCheck(id) {
+        if (registeredChecks[id]) {
+            return registeredChecks[id].complete;
+        }
+
+        return Promise.reject('no deferred check with id ' + id);
+    }
+
+    /**
+     * exposed for unit testing
+    **/
+    function _testRegisterCheck(check) {
```

Comment:
Yep it is, I didn't want to expose the private function registerCheck as it wasn't really part of the public interface for this module, so I created this almost identical function _testRegisterCheck as then it should be obvious to anyone that this has been exposed for testing purposes only. For the same reason I didn't want to expose checkList as this implies it can be accessed/modified by other modules.